lib/developer_portal/lib/cms/toolbar.rb

```diff
@@ -39,6 +39,30 @@ def draft?
 
   private
 
+  def validate_and_extract_cms_token
+    signature = params.delete(:signature)
+    return signature if signature.blank?
+
+    expires_at = Time.at(params.delete(:expires_at).to_i).utc
+    raise ActiveSupport::MessageVerifier::InvalidSignature unless expires_at > Time.now.utc
+
+    cms_token = site_account.settings.cms_token!
+    valid_signature = signature == CMS::Signature.generate(cms_token, expires_at)
+
+    raise ActiveSupport::MessageVerifier::InvalidSignature unless valid_signature
+
+    # We don't need the signature after processing, better remove it to avoid resending it with future redirections
+    request.query_parameters.delete(:expires_at)
+    request.query_parameters.delete(:signature)
+
+    cms_token
+  rescue StandardError
```

Comment:
what other error than `ActiveSupport::MessageVerifier::InvalidSignature` it might be?

Reply:
According to you: a big input error: https://github.com/3scale/porta/pull/3938#discussion_r1842984539